Mathlib/Tactic/DeriveToExpr.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+
+/-!
+# A `ToExpr` derive handler
+
+This module defines a `ToExpr` derive handler for inductive types. It supports mutually inductive
+types as well.
+
+It is originally based on the `Repr` derived handler from Lean 4 itself.
+
+This has a universe level limitation: we don't have the universe levels available for
+generating the expression, so we set all the universe level variables to zero (see
+where `levelZero` is used in this file).
```

Comment:
This is what the `reflected_univ` typeclass was for in Lean 3; but again, I think `Qq` might supercede it.

Reply:
Unfortunately the built-in ToExpr instances have this limitation already.  We could make our own type class.

Reply:
I think we did that in mathlib3 already; the instances are all in https://github.com/leanprover-community/mathlib3port/blob/master/Mathbin/Meta/Univs.lean, we just didn't port them.

Reply:
I mean we'd also need to add our own version of ToExpr.

Reply:
> I mean we'd also need to add our own version of ToExpr.
I don't follow, `ToExpr` is already universe-polymorphic

Reply:
https://github.com/leanprover/lean4/blob/8650804b025c9e69cd7f491eeb1ba988e6969367/src/Lean/ToExpr.lean#L74 for example

Reply:
Right, we would need to override the instances in core (as is done in the mathport file above), but there's no need to replace the typeclass itself.

Reply:
I added a `ToLevel` typeclass to give a bit of support for reflecting universe levels. To get around the "unused universe parameter" error it's set up where `toLevel (Sort u)` gives an expression for `u`.

Here's an example of what the derive handler gives for `List`:
```lean
mutual
  private def toExprList.{u} {α} [ToExpr α] [ToLevel (Sort u)] (x : @List.{u} α) : Expr :=
    match x with
    | @List.nil _ => mkAppN (Expr.const `List.nil [toLevel (Sort u)]) #[toTypeExpr α]
    | @List.cons _ a a' =>
      mkAppN (Expr.const `List.cons [toLevel (Sort u)]) #[toTypeExpr α, toExpr a, toExprList a']
end
instance {α} [ToExpr α] [ToLevel (Sort u)] : ToExpr (@List.{u} α) where
  toExpr := toExprList.{u}
  toTypeExpr := mkAppN (Expr.const `List [toLevel (Sort u)]) #[toTypeExpr α]
```

Reply:
What I meant was that the `ToExpr` type class has a type field, but no universe field.  You are right that we can just add an additional `ReflectedUniv`/`ToLevel` type class though.